modules/packages/LinearAlgebra.chpl

```diff
@@ -1244,49 +1244,38 @@ proc trace(A: [?D] ?eltType) {
 }
 
 private proc _lu (in A: [?Adom] ?eltType) {
-  const n = Adom.shape(0);
-  const LUDom = {1..n, 1..n};
+  const (n,m) = Adom.shape;
 
   // TODO: Reduce memory usage
-  var L, U, LU: [LUDom] eltType;
-
+  var L, LU: [Adom] eltType;
+  var U = A;
   var ipiv: [{1..n}] int = [i in {1..n}] i;
-
   var numSwap: int = 0;
 
   for i in 1..n {
 
-    var max = A[i,i], swaprow = i;
+    var max = U[i,i], swaprow = i;
     for row in (i+1)..n {
-      if (abs(A[row,i]) > abs(max)) {
-        max = A[row,i];
+      if (abs(U[row,i]) > abs(max)) {
+        max = U[row,i];
         swaprow = row;
       }
     }
     if (swaprow != i) {
-      A[i,..] <=> A[swaprow,..];
       L[i,..] <=> L[swaprow,..];
+      U[i,..] <=> U[swaprow,..];
       ipiv[i] <=> ipiv[swaprow];
       numSwap+=1;
     }
 
-    forall k in i..n {
-      var sum = + reduce (L[i,..] * U[..,k]);
-      U[i,k] = A[i,k] - sum;
-    }
-
-    L[i,i] = 1;
-
-    forall k in (i+1)..n {
-      var sum = + reduce (L[k,..] * U[..,i]);
-      L[k,i] = (A[k,i] - sum) / U[i,i];
+    if(max>0){
+      forall k in (i+1)..n {
+      L[k,i] = U[k,i] / U[i,i];
+      U[k,..] = U[k,..]-(L[k,i]*U[i,..]);
+      }
```

Comment:
I couldn't wrap my head around the implementation here, especially trying to answer "how's this equivalent to the previous implementation?". Could you provide some information as to the need for this change in the PR message? I am guessing it resolves one of the issues that we have had before, but I can't tell which and how?

Can you run the tests under https://github.com/chapel-lang/chapel/tree/master/test/library/packages/LinearAlgebra/performance/lu?

It'll require that you have lapack installed, and then, you can just do `start_test` under that directory. Even if you are sure that the logic here is right, it would be important to have the existing tests passing :)

Reply:
Hi @e-kayrakli ,
I believe the previous implementation of the procedure LU is not correct. It does not swap the rows of U while swapping the rows in A, L. Also while picking the pivot in the ith  iteration it uses the original matrix `A` instead of the elimination matrix `U`.
I'm referring to this part
``chpl 
for row in (i+1)..n {
      if (abs(A[row,i]) > abs(max)) {
        max = A[row,i];
        swaprow = row;
      }
    }
``
The method also does not handle the case when the pivot element is 0.

 I ran the performance tests locally on my Linux machine. The status of the performance results for the current implementation is same as the previous one.

Reply:
This solves the issue #16581 , by checking if the pivot is non-zero to proceed with elimination.

Reply:
Can you enter the data that you have in your tests and get the result from some other results and compare? I might be doing something wrong, but I am getting [this](https://www.wolframalpha.com/input/?i=%7Blu%2C+p%2C+c%7D+%3D+LUDecomposition%5B%7B1%2C+2%2C+3%7D%2C+%7B4%2C+5%2C+6%7D%2C+%7B8%2C+10%2C+12%7D%5D) for your first test case, for example.

Another thing of note, is that there's a race condition here, and your test fails. On repeated runs, I get different results for the last test case's U.

I also realized that I pointed you to the wrong test, try this one out: `test/library/packages/LinearAlgebra/performance/lu/lu-solve-test.chpl`, I see that it is failing with your PR.

You can also check out https://github.com/chapel-lang/chapel/pull/13516, which added LU to this library and has a quite a lot of discussion under it, in terms of correctness and stability of the implementation.

I do not see myself as an expert in linear algebra, but I remain unconvinced that this is the right implementation.

Reply:
I am also thinking that part of the confusion might be coming from the fact that the formal `A` has `in` intent, which means we copy the whole array upon call to this function. So, some of the changes you made (`var U = A`) makes me think that you have missed that `in` intent there.

Reply:
Hi @e-kayrakli ,
I think the reason for getting different lu decomposition from Wolfram is due the choice of pivot element during elimination. The algorithm I used for lu decomposition is Gaussian elimination with partial pivoting. I have been validating the test results for lu-decomposition with `scipy.linalg` which is based on the same algorithm, and gives the same decomposition result as the current implementation.

The reason we were getting different results for the last test case was because I had forgot to set a seed value for the procedure fillRandom :P. I'm fixing that now.

I think the test `test/library/packages/LinearAlgebra/performance/lu/lu-solve-test.chpl` fails on master as well. Please refer to issue #17074, for test logs.